include/swift/ABI/TaskGroup.h

```diff
@@ -0,0 +1,393 @@
+          str.append(" " + std::bitset<64>(status).to_string());
+          str.append(" }");
+          return str;
+        }
+
+        /// Initially there are no waiting and no pending tasks.
+        static const GroupStatus initial() {
+          return GroupStatus{0};
+        };
+    };
+
+    template<typename T>
+    class NaiveQueue {
+        std::queue<T> queue;
+
+    public:
+        NaiveQueue() = default;
+        NaiveQueue(const NaiveQueue<T> &) = delete ;
+        NaiveQueue& operator=(const NaiveQueue<T> &) = delete ;
+
+        NaiveQueue(NaiveQueue<T>&& other) {
+          queue = std::move(other.queue);
+        }
+
+        virtual ~NaiveQueue() { }
+
+        bool dequeue(T &output) {
+          if (queue.empty()) {
+            return false;
+          }
+          output = queue.front();
+          queue.pop();
+          return true;
+        }
+
+        void enqueue(const T item) {
+          queue.push(item);
+        }
+    };
+
+  private:
+
+//    // TODO: move to lockless via the status atomic
+    mutable std::mutex mutex;
+
+    /// Used for queue management, counting number of waiting and ready tasks
+    std::atomic<uint64_t> status;
+
+    /// TaskStatusRecord that is attached to the task running the group.
+    ///
+    /// Because we must remove it from the task as we exit/destroy the group,
+    /// we have to keep this pointer here so we know which record to remove then.
+    TaskGroupTaskStatusRecord* Record;
+
+    /// Queue containing completed tasks offered into this group.
+    ///
+    /// The low bits contain the status, the rest of the pointer is the
+    /// AsyncTask.
+    NaiveQueue<ReadyQueueItem> readyQueue;
+//     mpsc_queue_t<ReadyQueueItem> readyQueue; // TODO: can we get away with an MPSC queue here once actor executors land?
+
+    /// Single waiting `AsyncTask` currently waiting on `group.next()`,
+    /// or `nullptr` if no task is currently waiting.
+    std::atomic<AsyncTask*> waitQueue;
+
+    friend class AsyncTask;
+
+  public:
+    explicit TaskGroup(TaskGroupTaskStatusRecord* record)
+        : status(GroupStatus::initial().status),
+          Record(record),
+          readyQueue(),
+//          readyQueue(ReadyQueueItem::get(ReadyStatus::Empty, nullptr)),
+          waitQueue(nullptr) {}
+
+    /// Destroy the storage associated with the group.
+    void destroy(AsyncTask *task);
+
+    bool isEmpty() {
+      auto oldStatus = GroupStatus { status.load(std::memory_order_relaxed) };
+      return oldStatus.pendingTasks() == 0;
+    }
+
+    bool isCancelled() {
+      auto oldStatus = GroupStatus { status.load(std::memory_order_relaxed) };
+      return oldStatus.isCancelled();
+    }
+
+    TaskGroupTaskStatusRecord* getTaskRecord() const {
+      return Record;
+    }
+
+    /// Cancel the task group and all tasks within it.
+    ///
+    /// Returns `true` if this is the first time cancelling the group, false otherwise.
+    bool cancelAll(AsyncTask *task);
```

Comment:
the current task that is running the group.

I will document it.

----

It is passed so that we can:


```
  // cancel all existing tasks within the group
  swift_task_cancel_group_child_tasks(task, this);
```

If we make the group _be_ the record I think we won't need to pass the task 🤔